### ISSUE TITLE
Support OpenShift service account credentials for authentication

### ISSUE THREAD

**Issue by jtmulvey** [OpenLiberty/open-liberty#10022]
Open Liberty should be able to accept OpenShift service account credentials for authentication and role based access when configured to do so.  This would provide the ability for platform-supporting apps in OpenShift (such as Prometheus for monitoring) to authenticate with monitored Liberty servers with minimal required configuration.  When this is enabled, Liberty will map OpenShift project names to group names for the role-based access check.

UFO: https://ibm.box.com/s/wxrkaw020wt6ws9mxmlnv3opxrxikw7y

**Comment by atosak** [comment]
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - -a- - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [ ] POC Design / WAD Review Scheduled (David Chang) or N/A.
- [ ] POC Design / WAD Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [ ] Design / WAD Approval (Alasdair Nottingham) or N/A.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of the epic entitled "FAT Approval Test Summary".  Add and fill in the template as described here: https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [ ] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approved or No Design Approved label on the GitHub Epic.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Sam Wong - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
- [ ] **Demo** - (Liberty only - Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by clarkek123** [comment]
Serviceability Approval Comment - Please answer the following questions for serviceability approval:

WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

Yes. The WAD notes that many errors are already handled by the parent feature (socialLogin-1.0). New messages have been added to cover new errors that were not previously possible in config and errors with the service account token.

Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
a) What problem paths were tested and demonstrated?
**Configuration errors** 
Bad user validation API url - 401
Empty user validation API value - 401
Blank user validation API value - 401
Missing user validation API in config - defaults RC 200
Bad service account token 401 -CWWKS5383E
Missing service account token - 401 CWWKS5375E 
No bearer header in request - 401 CWWKS5375E
Blank realm name - defaults (realmSecurityName=https://openshift.apps.gawked.os.fyre.ibm.com:6443/libsrvacct2:satoken)

**Authorization RBAC errors**
application-bnd specifies realm that does not match that configured in realmName- 403 Authorization failed and message CWWKS9104A: Authorization failed for user libsrvacct2:satoken while invoking formlogin on /SimpleServlet. The user is not granted access to any of the required roles: [Employee, Manager].
application-bnd specifies a group that does not match the OpenShift project where service account token was created - 403 Authorization failed
application-bnd specifes a user that does not match the OpenShift service account token name - 403 Authorization failed

b) Who did you demo to?
**Test team and then
Provided test and development team with serviceability review on 2/10/20 with analysis of messages from both user perspective and system admin perspective on  (KC) **
The following issue was raised to address serviceability problems and all have now been fixed: 
https://github.com/OpenLiberty/open-liberty/issues/10821

c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
Kristen has run through the common error scenarios with build as of 02/18/20 and the agreed upon changes have been made to improve serviceability. Overall 1 message was improved and the other scenarios tested showed messages enable customers to resolve common problems without calling IBM support.

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature? 
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3? 

Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the following files and that the respective L2 / L3 teams know they are supporting it

**Comment by clarkek123** [comment]
@donbourne I have added the serviceability template for this feature.

**Comment by ayoho** [comment]
GA blog is being developed in https://github.com/OpenLiberty/open-liberty/issues/10952.

**Comment by chunlongliang-ibm** [comment]
@NottyCode please review and approve UFO, 
[UFOforEPIC10022.pptx](https://github.com/OpenLiberty/open-liberty/files/4222120/UFOforEPIC10022.pptx)

**Comment by donbourne** [comment]
@clarkek123 , will there be SVT for this feature?  I noticed that you left the questions about SVT unanswered.  The rest of what you described looks good.

**Comment by skasund** [comment]
L2 has requested a live STE presentation for this feature.  The STE templates can be found at the links below.  You can use either one to create the education.  

Slide Template:  https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Github Template:  https://pages.github.ibm.com/WASL3/site/STE/about

Please upload the completed slides to the same STE Archive BOX folder, and I will provide the STE approval.  The live presentation can be scheduled with the L2 team at a later date.  Thanks!

**Comment by clarkek123** [comment]
@donbourne Rumana Haque will be looking at this from an SVT standpoint. I can update the serviceability section once I hear back from her.

**Comment by ayoho** [comment]
@skasund I've just uploaded STE-Feature-10022-OpenShift-service-accounts.ppt to the Box folder you provided.

**Comment by ayoho** [comment]
Opened https://github.com/OpenLiberty/docs/issues/1118 to update the Open Liberty docs with examples on how to use this feature.

**Comment by chirp1** [comment]
Adam supplied the info for ID to document in this issue: https://github.com/OpenLiberty/docs/issues/1118 As I mentioned to Adam, ID is working down a backlog, so the writers might not get to this issue until after 20.0.0.3. However, Adam will have a blog post with meaningful info for customers. Approving.

**Comment by donbourne** [comment]
> @donbourne Rumana Haque will be looking at this from an SVT standpoint. I can update the serviceability section once I hear back from her.

Thanks @clarkek123 -- let me know when that's done and assuming Rumana is happy with the serviceability I can approve.

**Comment by skasund** [comment]
@ayoho  Thanks for the STE slides.  I'll have the L2 team reach out to set up the live presentation.  I've approved the feature.

**Comment by gscottj** [comment]
This feature does not add any new or change existing user interfaces. No addition accessibility verification testing required.

**Comment by clarkek123** [comment]
@donbourne, I have serviceability approval from Rumana of SVT and have updated the Serviceability section to indicate this.

**Comment by rumanaHaque** [comment]
I am the SVT Focal Point for this Feature. I have reviewed the FVT test cases for this Feature, including the configuration. The FVT test involved taking a Liberty server configured for the new feature, dockerizing it and deploying in OpenShift, and then using an OpenShift service account token in a Bearer token request header in Postman to try to access a protected servlet in Liberty.
This would be very similar to any test we would attempt for SVT. Looking at FVT test configuration, the results of the test, as well as the Serviceability messages, I think the FVT tests provide sufficient coverage, so no explicit SVT is required. I discussed this with Chunlong, and he agreed as well.
Based on this, I will be accepting the SVT focal point approval for this Feature.

**Comment by chunlongliang-ibm** [comment]
[UFOforEPIC10022-March03.pdf](https://github.com/OpenLiberty/open-liberty/files/4284049/UFOforEPIC10022-March03.pdf)

**Comment by tevans78** [comment]
Removed the `target:ga` label until this feature has design approval.
